homeassistant/components/esphome/sensor.py

```diff
@@ -71,9 +74,84 @@ async def async_setup_entry(
 )
 
 
-class EsphomeSensor(EsphomeEntity[SensorInfo, SensorState], SensorEntity):
+class EsphomeSensor(
+    EsphomeEntity[SensorInfo, SensorState], SensorEntity, RestoreEntity
+):
     """A sensor implementation for esphome."""
 
+    async def async_added_to_hass(self) -> None:
+        """Register callbacks."""
+        await super().async_added_to_hass()
+
+        if self._static_info.last_reset_type != LastResetType.AUTO:
+            return
+
+        # Logic to restore old state for last_reset_type AUTO:
+        last_state = await self.async_get_last_state()
+        if last_state is None:
+            return
+
+        if "last_reset" in last_state.attributes:
+            self._attr_last_reset = datetime.fromisoformat(
+                last_state.attributes["last_reset"]
+            )
+
+        try:
+            self._old_state = float(last_state.state)
+        except ValueError:
+            pass
+
+    _old_state: float | None = None
+
+    @callback
+    def _on_state_update(self) -> None:
+        """Check last_reset when new state arrives."""
+        if self._static_info.last_reset_type == LastResetType.NEVER:
+            self._attr_last_reset = datetime.min
+
+        if self._static_info.last_reset_type != LastResetType.AUTO:
+            super()._on_state_update()
+            return
+
+        # Last reset type AUTO logic for the last_reset property
+        # In this mode we automatically determine if an accumulator reset
+        # has taken place.
+        # We compare the last valid value (_old_state) with the new one.
+        # If the value has reset to 0 or has significantly reduced we say
+        # it has reset.
+        new_state: float | None = None
+        state = cast("str | None", self.state)
+        if state is not None:
+            try:
+                new_state = float(state)
+            except ValueError:
+                pass
+
+        did_reset = False
+        if new_state is None:
+            # New state is not a float - we'll detect the reset once we get valid data again
+            did_reset = False
+        elif self._old_state is None:
+            # First measurement we ever got for this sensor, always a reset
+            did_reset = True
+        elif new_state == 0:
+            # don't set reset if both old and new are 0
+            # we would already have detected the reset on the last state
+            did_reset = self._old_state != 0
+        elif new_state < self._old_state / 2:
+            # Significantly decreased new state
+            did_reset = True
+
+        # Set last_reset to now if we detected a reset
+        if did_reset:
+            self._attr_last_reset = datetime.now()
```

Comment:
Times in the state machine must always be UTC time. Please use `util.dt.utcnow`.